File: Algo-1/week3/7-Range-Minimum-Query/rmq.cpp
#include <iostream>
#include <climits>
#include <cstring>
#include <string>
#include <stdio.h>

using namespace std;

class RMQ {
public:
	int size;
	int* tree;

	RMQ(int size) {
		int k = 1;
		// get the biggest power of two, greater or euql to size
		while (k < size) { 
			k *= 2;
		}

		this->size = k;
		k *= 2;
		tree = new int[k];
		for (int i = 0; i < k; i++) {
			tree[i] = INT_MAX;
		}
	}

	~RMQ() {
		//delete[] tree;
	}

	// sets the value at index
	void set(int index, int value) {
		index += size;
		tree[index] = value;
		index /= 2;

		while (index > 0) {
			int leftChild = index * 2;
			int rightChild = index * 2 + 1;

			tree[index] = (tree[leftChild] > tree[rightChild] ? tree[rightChild] : tree[leftChild]);
		
			index /= 2;
		}
	}

	// returns the minimum value in a range
	int min(int startIndex, int endIndex) {
		startIndex += size;
		endIndex += size;

		int min = INT_MAX;

		if (endIndex == startIndex) {
			return min > tree[endIndex] ? tree[endIndex] : min;
		}

		while (endIndex > startIndex) {
			if (endIndex % 2 == 0) {
				min = (min > tree[endIndex] ? tree[endIndex] : min);
				endIndex /= 2;
				endIndex--;
			}
			else {
				endIndex /= 2;
				min = (min > tree[endIndex] ? tree[endIndex] : min);
			}

			if (startIndex % 2) {
				min = (min > tree[startIndex] ? tree[startIndex] : min);
				startIndex /= 2;
				startIndex++;

				if (startIndex == endIndex) {
					min = (min > tree[startIndex] ? tree[startIndex] : min);
					break;
				}
			}
			else {
				startIndex /= 2;
				min = (min > tree[startIndex] ? tree[startIndex] : min);
			}

		}

		return min;
	}

};

int main() {
	int n,q;
  	scanf("%d%d",&n,&q);

  	RMQ rmq(n);

  	for(int i=0;i<n;i++) {
    		int temp;
    		scanf("%d",&temp);
    		rmq.set(i,temp);
  	}

  	for(int i=0;i<q;i++) {
    		char s[3];
    		int a,b;
    		scanf("%s%d%d",s,&a,&b);
    		if (strcmp(s,"min")==0) // if s is "min"
      			printf("%d\n",rmq.min(a,b));
    		else {
      			rmq.set(a,b);
    		}		
  	}

  	return 0;
}

File: Algo-1/week3/7-Range-Minimum-Query/rmq.py
import math


class RMQ:

    MAX_VALUE = 2 ** 31

    def __init__(self, size):
        self.level = RMQ.determine_level(size)
        self.values = [RMQ.MAX_VALUE] * (2 ** (self.level + 1))

    @staticmethod
    def determine_level(elements_count):
        a = 1
        level = 0

        while a < elements_count:
            a *= 2
            level += 1

        return level

    @staticmethod
    def find_position(index, level):
        return 2 ** level + index - 1

    # sets the value at index
    def set(self, index, value):
        position = RMQ.find_position(index, self.level)

        self.values[position] = value
        position = (position - 1) // 2

        while True:
            self.values[position] = min(self.values[2 * position + 1], self.values[2 * position + 2])
            parent = (position - 1) // 2
            if position == 0:
                break
            else:
                position = parent

    # returns the minimum value in a range
    def minimum(self, start_index, end_index):
        level = self.level
        result = self.values[RMQ.find_position(start_index, level)]

        while start_index <= end_index:
            result = min(result, self.values[RMQ.find_position(start_index, level)], self.values[RMQ.find_position(end_index, level)])

            if end_index % 2 == 0:
                end_index = (end_index // 2) - 1
            else:
                end_index //= 2

            if start_index % 2 == 0:
                start_index //= 2
            else:
                start_index = (start_index // 2) + 1

            level -= 1

        return result


def main():
    line = input().split()
    N = int(line[0])
    K = int(line[1])

    values = input().split()
    rmq = RMQ(N)

    for index in range(len(values)):
        rmq.set(index, int(values[index]))

    while K != 0:
        command = input().split()
        if command[0] == 'set':
            rmq.set(int(command[1]), int(command[2]))
        else:
            print(rmq.minimum(int(command[1]), int(command[2])))
        K -= 1


if __name__ == '__main__':
    main()
